### DISCUSSION TITLE
Graceful Cancellation & ExitedCommandEvent?

### DISCUSSION THREAD

**Discussion by mikoskinen** [Tyrrrz/CliWrap#224] [Q&A]
Hi,

We've been using CliWrap for a while now and it has been working nicely. The only problem has been with some of the processes not closing when the main app is killed so we're looking if the Graceful Cancellation -feature could help us.

We're running the executables using:

`await runCmd.Observe(Console.OutputEncoding, Console.OutputEncoding, forcefulCts.Token, flow.CancellationToken).ForEachAsync`

Based on the preliminary testing we noticed that if we use the graceful CancellationToken to exit the process, we are not receiving the ExitedCommandEvent.

Is this by design? Or should we still receive ExitedCommandEvent when using Graceful Cancellation?

**Comment by Tyrrrz** [comment]
Hi.

This is indeed by design. When a command execution gets canceled, an exception gets thrown, so it never really reaches the point where it checks the process exit code, etc. In case with `IObservable`, you should get the error that corresponds to the cancellation as one of the events.

https://github.com/Tyrrrz/CliWrap/blob/c4d5f948c57d08f5e3ec17edcbda91b6b943fdd3/CliWrap/EventStream/PushEventStreamCommandExtensions.cs#L68-L71
